# CS246 - Colab 5
## PageRank

### Setup

First of all, we authenticate a Google Drive client to download the dataset we will be processing in this Colab.

**Make sure to follow the interactive instructions.**

In [1]:
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [2]:
id='1EoolSK32_U74I4FeLox88iuUB_SUUYsI'
downloaded = drive.CreateFile({'id': id})
downloaded.GetContentFile('web-Stanford.txt')

If you executed the cells above, you should be able to see the dataset we will use for this Colab under the "Files" tab on the left panel.

Next, we import some of the common libraries needed for our task.

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

### Data Loading

For this Colab we will be using [NetworkX](https://networkx.github.io), a Python package for the creation, manipulation, and study of the structure, dynamics, and functions of complex networks.

The dataset we will analyze is a snapshot of the Web Graph centered around [stanford.edu](https://stanford.edu), collected in 2002. Nodes represent pages from Stanford University (stanford.edu) and directed edges represent hyperlinks between them. [[More Info]](http://snap.stanford.edu/data/web-Stanford.html)

In [4]:
import networkx as nx

G = nx.read_edgelist('web-Stanford.txt', create_using=nx.DiGraph)

In [5]:
print(nx.info(G))

Name: 
Type: DiGraph
Number of nodes: 281903
Number of edges: 2312497
Average in degree:   8.2032
Average out degree:   8.2032


### Your Task

To begin with, let's simplify our analysis by ignoring the dangling nodes and the disconnected components in the original graph.

Use NetworkX to identify the **largest** weakly connected component in the ```G``` graph.  From now on, use this connected component for all the following tasks.

In [6]:
G_largest = max(nx.weakly_connected_components(G),key=len)

In [7]:
C = nx.subgraph(G, G_largest)
print(nx.info(C))

Name: 
Type: DiGraph
Number of nodes: 255265
Number of edges: 2234572
Average in degree:   8.7539
Average out degree:   8.7539


Compute the PageRank vector, using the default parameters in NetworkX: [https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.link_analysis.pagerank_alg.pagerank.html#networkx.algorithms.link_analysis.pagerank_alg.pageranky](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.link_analysis.pagerank_alg.pagerank.html#networkx.algorithms.link_analysis.pagerank_alg.pagerank)

In [8]:
pr_orig = nx.pagerank(C)

In 1999, Barabási and Albert proposed an elegant mathematical model which can generate graphs with topological properties similar to the Web Graph (also called Scale-free Networks).

If you complete the steps below, you should obtain some empirical evidence that the Random Graph model is inferior compared to the Barabási–Albert model when it comes to generating a graph resembling the World Wide Web!

As such, we will use two different graph generator methods, and then we will test how well they approximate the Web Graph structure by means of comparing the respective PageRank vectors. [[NetworkX Graph generators]](https://networkx.github.io/documentation/stable/reference/generators.html#)

Using for both methods ```seed = 1```, generate:


1.   a random graph (with the fast method), setting ```n``` equal to the number of nodes in the original connected component, and ```p = 0.00008```
2.   a Barabasi-Albert graph (with the standard method), setting ```n``` equal to the number of nodes in the original connected component, and finding the right ***integer*** value for ```m``` such as the resulting number of edges **approximates by excess** the number of edges in the original connected component

and compute the PageRank vectors for both graphs.


In [27]:
fast_graph = nx.fast_gnp_random_graph(n = 255265, p = 0.00008, seed = 1)

In [28]:
fast_graph.number_of_edges()

2606386

In [29]:
pr_random = nx.pagerank(fast_graph)

In [30]:
max(pr_random, key=pr_random.get)

150787

In [31]:
BA_graph = nx.barabasi_albert_graph(255265, m = 9, seed = 1)

In [32]:
BA_graph.number_of_edges()

2297304

In [33]:
pr_BA = nx.pagerank(BA_graph)

In [34]:
max(pr_BA, key = pr_BA.get)

11

In [35]:
pr_BA[11]

0.00042292470661201644

Compare the PageRank vectors obtained on the generated graphs with the PageRank vector you computed on the original connected component.
**Sort** the components of each vector by value, and use cosine similarity as similarity measure. 

Feel free to use any implementation of the cosine similarity available in third-party libraries, or implement your own with ```numpy```.

In [36]:
ORIG = sorted(list(pr_orig.values()))

In [37]:
RANDOM = sorted(list(pr_random.values()))

In [38]:
BA = sorted(list(pr_BA.values()))

In [39]:
from scipy import spatial
result1 = 1 - spatial.distance.cosine(ORIG, RANDOM)
result2 = 1 - spatial.distance.cosine(ORIG, BA)

In [42]:
result1

0.10395564703964189

In [43]:
result2

0.6612451256426742

Once you have working code for each cell above, **head over to Gradescope, read carefully the questions, and submit your solution for this Colab**!

## Results:
----------------
________________________

*   Q2.1 : 255265
*   Q2.2 : 8.754
----------------------
*   Q3.1 : 2606386
*   Q3.2 : 150787
*   Q3.3 : 2297304
*   Q3.4 : (10^-4, 10^-3)
---------
*   Q4.1 : 0.104
*   Q4.2 : 0.661

----------------
________________________
